File: Problems/0261. Graph Valid Tree/solution2.js
// valid tree: n nodes labeled 0..n-1, edges as [[u,v], ...]
function validTree(n, edges) {
  // Necessary condition for a tree
  if (edges.length !== n - 1) return false;

  // Build undirected adjacency list
  const graph = Array.from({ length: n }, () => []);
  for (const [u, v] of edges) {
    graph[u].push(v);
    graph[v].push(u);
  }

  const visited = new Array(n).fill(false);

  // DFS with parent tracking to detect cycles
  function hasCycle(node, parent) {
    if (visited[node]) return true; // revisit => cycle
    visited[node] = true;

    for (const nei of graph[node]) {
      if (nei === parent) continue; // skip the edge we came from
      if (hasCycle(nei, node)) return true;
    }
    return false;
  }

  if (hasCycle(0, -1)) return false;

  // Connectivity check: all nodes must be visited
  for (const v of visited) if (!v) return false;

  return true;
}

File: Problems/0261. Graph Valid Tree/solution2.py
class Solution:
    def validTree(self, n: int, edges: List[List[int]]) -> bool:
        adjList = defaultdict(list)
        visited = set()

        for node, neighbor in edges:
            adjList[node].append(neighbor)
            adjList[neighbor].append(node)

        def dfs(node, prevNode, path):   
            if node in path:
                return False
            
            path.add(node)
            
            for neighbor in adjList[node]:
                if neighbor != prevNode:
                    if not dfs(neighbor, node, path):
                        return False

            return True
        
        return dfs(0, -1, visited) and len(visited) == n
